File: examples/mobile/android/app/src/main/java/org/ospreylang/issueinbox/NativeRenderer.kt
package org.ospreylang.issueinbox

import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.text.Editable
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.TextWatcher
import android.text.method.LinkMovementMethod
import android.text.style.StyleSpan
import android.text.style.TypefaceSpan
import android.text.style.URLSpan
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.widget.*
import org.json.JSONArray
import org.json.JSONObject
import java.util.WeakHashMap

// Generic native projection of the tree built by Osprey Ui. [MOBILE-REACTIVE-UI]
internal class NativeRenderer(private val context: Context, private val dispatch: (JSONObject) -> Unit) {
    private val nodes = WeakHashMap<View, JSONObject>()
    private var rendering = false
    private val ink = Color.rgb(232, 239, 250)
    private val muted = Color.rgb(159, 177, 201)
    private val accent = Color.rgb(70, 223, 194)
    private fun dp(value: Int) = (value * context.resources.displayMetrics.density).toInt()

    fun update(container: FrameLayout, node: JSONObject) {
        rendering = true
        try {
            val tree = render(node, container.getChildAt(0))
            if (tree.parent == null) { container.removeAllViews(); container.addView(tree) }
        } finally { rendering = false }
    }

    private fun render(node: JSONObject, previous: View?): View {
        val same = previous != null && nodes[previous]?.optString("kind") == node.getString("kind") &&
            nodes[previous]?.optString("id") == node.optString("id")
        val view = if (same) previous!! else create(node.getString("kind"))
        nodes[view] = node
        when (view) {
            is EditText -> input(view, node)
            is TextView -> text(view, node)
            is ViewGroup -> children(view, node)
        }
        style(view, node.optString("style"))
        return view
    }

    private fun create(kind: String): View = when (kind) {
        "column", "row" -> LinearLayout(context).apply {
            orientation = if (kind == "row") LinearLayout.HORIZONTAL else LinearLayout.VERTICAL
            gravity = Gravity.CENTER_VERTICAL
        }
        "scroll" -> ScrollView(context).apply { isFillViewport = true }
        "text", "rich" -> TextView(context)
        "button", "link" -> Button(context).apply { isAllCaps = false }
        "input" -> newInput()
        "divider" -> View(context).apply { setBackgroundColor(muted); minimumHeight = dp(1) }
        else -> error("Unknown Osprey UI node kind: $kind")
    }

    private fun children(parent: ViewGroup, node: JSONObject) {
        val children = node.optJSONArray("children") ?: return
        for (index in 0 until children.length()) {
            val childNode = children.getJSONObject(index)
            val previous = parent.getChildAt(index)
            val child = render(childNode, previous)
            if (child !== previous) {
                if (previous != null) parent.removeViewAt(index)
                parent.addView(child, index)
            }
            layout(parent, child, childNode)
        }
        while (parent.childCount > children.length()) parent.removeViewAt(parent.childCount - 1)
    }

    private fun layout(parent: ViewGroup, child: View, node: JSONObject) {
        val row = parent is LinearLayout && parent.orientation == LinearLayout.HORIZONTAL
        val flexible = row && node.getString("kind") in listOf("text", "rich")
        val width = if (flexible) 0 else if (row) -2 else -1
        child.layoutParams = if (parent is LinearLayout) LinearLayout.LayoutParams(width, -2).apply {
            if (flexible) weight = 1f
            setMargins(0, dp(4), if (row) dp(8) else 0, dp(4))
        } else FrameLayout.LayoutParams(-1, -2)
    }

    private fun text(view: TextView, node: JSONObject) {
        val rich = node.getString("kind") == "rich"
        view.text = if (rich) spans(node.optJSONArray("spans") ?: JSONArray()) else node.optString("text")
        if (rich) view.movementMethod = LinkMovementMethod.getInstance()
        view.setOnClickListener(if (node.has("event")) View.OnClickListener { dispatch(node.getJSONObject("event")) }
            else if (node.getString("kind") == "link") View.OnClickListener {
                val uri = Uri.parse(node.getString("url"))
                require(uri.scheme == "https") { "Only HTTPS links are supported" }
                context.startActivity(Intent(Intent.ACTION_VIEW, uri))
            } else null)
    }

    // Implements [MOBILE-MARKDOWN]: spans carry Osprey's emphasis, code and
    // HTTPS link decisions; Android only applies the platform presentation.
    private fun spans(spans: JSONArray): CharSequence {
        val builder = SpannableStringBuilder()
        for (index in 0 until spans.length()) {
            val span = spans.getJSONObject(index)
            val start = builder.length
            builder.append(span.optString("text"))
            val style: Any? = when (span.optString("style")) {
                "bold" -> StyleSpan(Typeface.BOLD)
                "italic" -> StyleSpan(Typeface.ITALIC)
                "code" -> TypefaceSpan("monospace")
                else -> null
            }
            if (style != null) builder.setSpan(style, start, builder.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            val url = span.optString("url")
            if (url.startsWith("https://")) builder.setSpan(URLSpan(url), start, builder.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
        }
        return builder
    }

    private fun newInput() = EditText(context).apply {
        setSingleLine(true)
        inputType = android.text.InputType.TYPE_CLASS_TEXT or android.text.InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS
        addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                val node = nodes[this@apply] ?: return
                if (!rendering && !node.optBoolean("submit")) changed(this@apply, node)
            }
            override fun afterTextChanged(s: Editable?) {}
        })
    }

    private fun changed(view: EditText, node: JSONObject) {
        dispatch(JSONObject(node.getJSONObject("event").toString()).put("value", view.text.toString()))
    }

    private fun input(view: EditText, node: JSONObject) {
        view.hint = node.optString("placeholder")
        view.contentDescription = node.optString("id")
        val value = node.optString("value")
        if (!view.hasFocus() && view.text.toString() != value) view.setText(value)
        view.imeOptions = EditorInfo.IME_ACTION_DONE
        view.setOnEditorActionListener { _, action, key ->
            if (node.optBoolean("submit") && (action == EditorInfo.IME_ACTION_DONE ||
                    (key?.keyCode == android.view.KeyEvent.KEYCODE_ENTER && key.action == android.view.KeyEvent.ACTION_UP))) {
                changed(view, node); true
            } else false
        }
    }

    private fun style(view: View, name: String) {
        if (view is TextView) {
            view.setTextColor(when (name) { "accent" -> accent; "caption", "bullet", "quote" -> muted; "error" -> Color.rgb(255, 143, 143); "primary" -> Color.rgb(12, 30, 36); else -> ink })
            view.textSize = when (name) { "hero" -> 36f; "heading" -> 22f; "title" -> 19f; "subheading" -> 18f; "caption", "accent" -> 13f; "code" -> 14f; else -> 16f }
            view.setTypeface(if (name == "code") Typeface.MONOSPACE else null,
                if (name in listOf("hero", "title", "accent", "heading", "subheading")) Typeface.BOLD else Typeface.NORMAL)
            if (view is EditText) view.setHintTextColor(muted)
        }
        if (name in listOf("screen", "card", "primary", "secondary", "code")) {
            view.background = GradientDrawable().apply {
                setColor(when (name) { "screen" -> Color.rgb(12, 20, 34); "primary" -> accent; "code" -> Color.rgb(18, 28, 44); else -> Color.rgb(25, 39, 59) })
                cornerRadius = if (name == "screen") 0f else dp(14).toFloat()
                if (name == "secondary") setStroke(dp(1), Color.rgb(53, 74, 100))
            }
            val padding = dp(when (name) { "screen" -> 20; "code" -> 12; else -> 14 })
            view.setPadding(padding, padding, padding, padding)
        } else if (name == "quote") {
            view.background = null
            view.setPadding(dp(12), 0, 0, 0)
        } else if (view is ViewGroup) {
            view.background = null
            view.setPadding(0, 0, 0, 0)
        }
    }
}

File: examples/mobile/ios/IssueInbox/NativeRenderer.swift
import SwiftUI

// Implements [MOBILE-REACTIVE-UI]. All labels, layout and events come from Osprey.
struct InboxNode: Decodable {
    let kind: String
    var id: String?
    var text: String?
    var style: String?
    var value: String?
    var placeholder: String?
    var url: String?
    var submit: Bool?
    var event: [String: String]?
    var children: [InboxNode]?
    var spans: [InboxSpan]?
}

// One run of a `rich` node: Osprey chose the style and link. [MOBILE-MARKDOWN]
struct InboxSpan: Decodable {
    let text: String
    let style: String
    var url: String?
}

struct NativeRenderer: View {
    let node: InboxNode
    let send: ([String: Any]) -> Void

    var body: some View {
        rendered.modifier(NodeStyle(style: node.style ?? ""))
            .accessibilityIdentifier(node.id ?? "")
    }

    private var children: some View {
        ForEach(Array((node.children ?? []).enumerated()), id: \.offset) { _, child in
            NativeRenderer(node: child, send: send)
        }
    }

    private var rendered: AnyView {
        switch node.kind {
        case "column": return AnyView(VStack(alignment: .leading, spacing: 14) { children })
        case "row": return AnyView(HStack(alignment: node.style == "item" ? .top : .center,
                                          spacing: node.style == "item" ? 8 : 14) { children })
        case "scroll": return AnyView(ScrollView { children })
        case "text": return AnyView(Text(node.text ?? "").fixedSize(horizontal: false, vertical: true))
        case "rich": return AnyView(Text(richText).fixedSize(horizontal: false, vertical: true))
        case "button": return AnyView(Button(node.text ?? "") { send(node.event ?? [:]) })
        case "input": return AnyView(NativeInput(node: node, send: send))
        case "link": return link
        case "divider": return AnyView(Divider())
        default: return AnyView(Text("Unsupported native view: \(node.kind)").foregroundColor(.red))
        }
    }

    // Implements [MOBILE-MARKDOWN]: spans carry Osprey's emphasis, code and
    // HTTPS link decisions; SwiftUI only applies the platform presentation.
    private var richText: AttributedString {
        (node.spans ?? []).reduce(into: AttributedString()) { result, span in
            var run = AttributedString(span.text)
            switch span.style {
            case "bold": run.inlinePresentationIntent = .stronglyEmphasized
            case "italic": run.inlinePresentationIntent = .emphasized
            case "code": run.inlinePresentationIntent = .code
            default: break
            }
            if let raw = span.url, let url = URL(string: raw), url.scheme == "https" { run.link = url }
            result.append(run)
        }
    }

    private var link: AnyView {
        guard let raw = node.url, let url = URL(string: raw), url.scheme == "https" else {
            return AnyView(Text("Invalid link").foregroundColor(.red))
        }
        return AnyView(Link(node.text ?? raw, destination: url))
    }
}

private struct NativeInput: View {
    let node: InboxNode
    let send: ([String: Any]) -> Void
    @State private var value: String
    @FocusState private var focused: Bool

    init(node: InboxNode, send: @escaping ([String: Any]) -> Void) {
        self.node = node
        self.send = send
        _value = State(initialValue: node.value ?? "")
    }

    var body: some View {
        TextField(node.placeholder ?? "", text: $value)
            .textInputAutocapitalization(.never).autocorrectionDisabled()
            .submitLabel(node.submit == true ? .go : .search).focused($focused)
            .padding(13).background(Color(.tertiarySystemGroupedBackground), in: RoundedRectangle(cornerRadius: 12))
            .onChange(of: value) { _ in if node.submit != true { dispatch() } }
            .onChange(of: node.value) { next in if !focused { value = next ?? "" } }
            .onSubmit { dispatch(); focused = false }
    }

    private func dispatch() {
        var event: [String: Any] = node.event ?? [:]
        event["value"] = value
        send(event)
    }
}

private struct NodeStyle: ViewModifier {
    let style: String
    @Environment(\.colorScheme) private var scheme
    private let fill = Color(red: 0.29, green: 0.25, blue: 0.81)
    private var accent: Color {
        scheme == .dark ? Color(red: 0.68, green: 0.65, blue: 1) : fill
    }

    @ViewBuilder func body(content: Content) -> some View {
        switch style {
        case "screen": content.background(Color(.systemGroupedBackground)).tint(accent)
        case "page": content.padding(20).frame(maxWidth: 720)
        case "header": content.padding(.vertical, 14).frame(maxWidth: .infinity, alignment: .leading)
        case "card": content.padding(18).frame(maxWidth: .infinity, alignment: .leading)
                .background(Color(.secondarySystemGroupedBackground), in: RoundedRectangle(cornerRadius: 20))
        case "hero": content.font(.system(size: 36, weight: .bold, design: .rounded))
        case "title": content.font(.headline)
        case "caption": content.font(.caption).foregroundColor(.secondary)
        case "accent": content.font(.caption.weight(.bold)).foregroundColor(accent)
        case "error": content.font(.callout).foregroundColor(.red)
        case "heading": content.font(.title3.weight(.semibold))
        case "subheading": content.font(.headline)
        case "code": content.font(.system(.footnote, design: .monospaced)).padding(12).frame(maxWidth: .infinity, alignment: .leading)
                .background(Color(.tertiarySystemGroupedBackground), in: RoundedRectangle(cornerRadius: 10))
        case "quote": content.padding(.leading, 12).foregroundColor(.secondary)
                .overlay(alignment: .leading) { RoundedRectangle(cornerRadius: 2).fill(accent.opacity(0.6)).frame(width: 3) }
        case "bullet": content.foregroundColor(.secondary)
        case "item": content.frame(maxWidth: .infinity, alignment: .leading)
        case "primary": content.font(.subheadline.weight(.semibold)).padding(.horizontal, 16).padding(.vertical, 10)
                .foregroundColor(.white).background(fill, in: Capsule())
        case "secondary": content.font(.subheadline.weight(.medium)).padding(.horizontal, 12).padding(.vertical, 10)
                .foregroundColor(accent).background(accent.opacity(0.09), in: Capsule())
        default: content
        }
    }
}
